File: ruby/lib/pyrosql/codec.rb
# frozen_string_literal: true

require "extlz4"

module PyroSQL
  # PWire binary protocol codec for PyroSQL.
  # Header format: 1 byte type + 4 bytes little-endian length + payload.
  module Codec
    # Client -> server message types
    MSG_QUERY      = 0x01
    MSG_PREPARE    = 0x02
    MSG_EXECUTE    = 0x03
    MSG_CLOSE      = 0x04
    MSG_PING       = 0x05
    MSG_AUTH       = 0x06
    MSG_COMPRESSED = 0x10
    MSG_QUIT       = 0xFF

    # Capability flags
    CAP_LZ4 = 0x01
    COMPRESSION_THRESHOLD = 8 * 1024

    # Server -> client response types
    RESP_RESULT_SET = 0x01
    RESP_OK         = 0x02
    RESP_ERROR      = 0x03
    RESP_PONG       = 0x04
    RESP_READY      = 0x05

    # Value type tags
    TYPE_NULL  = 0
    TYPE_I64   = 1
    TYPE_F64   = 2
    TYPE_TEXT  = 3
    TYPE_BOOL  = 4
    TYPE_BYTES = 5

    HEADER_SIZE = 5

    class << self
      # Build a framed PWire message.
      def frame(type, payload = "".b)
        payload = payload.b if payload.is_a?(String) && payload.encoding != Encoding::BINARY
        header = [type, payload.bytesize].pack("CV")
        header + payload
      end

      def encode_auth(user, password)
        user_bytes = user.encode("UTF-8").b
        pass_bytes = password.encode("UTF-8").b
        payload = [user_bytes.bytesize].pack("C") + user_bytes +
                  [pass_bytes.bytesize].pack("C") + pass_bytes
        frame(MSG_AUTH, payload)
      end

      # Build an AUTH frame with capability byte for LZ4 negotiation.
      def encode_auth_with_caps(user, password, caps)
        user_bytes = user.encode("UTF-8").b
        pass_bytes = password.encode("UTF-8").b
        payload = [user_bytes.bytesize].pack("C") + user_bytes +
                  [pass_bytes.bytesize].pack("C") + pass_bytes +
                  [caps].pack("C")
        frame(MSG_AUTH, payload)
      end

      # Compress a frame payload with LZ4 if it exceeds the threshold.
      # Returns a MSG_COMPRESSED frame or the original frame.
      # Uses lz4_flex-compatible format: [u32 LE original size][raw lz4 block].
      def compress_frame(type, payload)
        payload = payload.b if payload.is_a?(String) && payload.encoding != Encoding::BINARY
        return frame(type, payload) if payload.bytesize <= COMPRESSION_THRESHOLD

        raw_compressed = LZ4.block_encode(payload)
        # lz4_flex format: prepend u32 LE original size
        lz4_data = [payload.bytesize].pack("V") + raw_compressed

        ratio = lz4_data.bytesize.to_f / payload.bytesize
        return frame(type, payload) if ratio > 0.9

        # Inner: [original_type: u8][uncompressed_length: u32 LE][lz4_data]
        inner = [type].pack("C") + [payload.bytesize].pack("V") + lz4_data
        frame(MSG_COMPRESSED, inner)
      end

      # Decompress a MSG_COMPRESSED frame payload.
      # Returns [original_type, decompressed_payload].
      # lz4_data is in lz4_flex format: [u32 LE original size][raw lz4 block].
      def decompress_frame(payload)
        raise PyroSQL::ConnectionError, "Compressed payload too short" if payload.bytesize < 5

        original_type = payload.getbyte(0)
        uncompressed_len = payload[1, 4].unpack1("V")
        lz4_data = payload[5..]

        # lz4_data starts with [u32 LE original size] from lz4_flex, skip it
        if lz4_data.bytesize >= 4
          raw_block = lz4_data[4..]
          decompressed = LZ4.block_decode(raw_block, uncompressed_len)
        else
          raise PyroSQL::ConnectionError, "LZ4 data too short"
        end

        [original_type, decompressed]
      end

      def encode_query(sql)
        frame(MSG_QUERY, sql.encode("UTF-8").b)
      end

      def encode_prepare(sql)
        frame(MSG_PREPARE, sql.encode("UTF-8").b)
      end

      def encode_execute(handle, params)
        payload = [handle].pack("V") + [params.size].pack("v")
        params.each do |p|
          p_bytes = p.to_s.encode("UTF-8").b
          payload += [p_bytes.bytesize].pack("v") + p_bytes
        end
        frame(MSG_EXECUTE, payload)
      end

      def encode_close(handle)
        frame(MSG_CLOSE, [handle].pack("V"))
      end

      def encode_ping
        frame(MSG_PING)
      end

      def encode_quit
        frame(MSG_QUIT)
      end

      # Read a complete frame from an IO stream.
      # Returns [type, payload].
      # Transparently decompresses MSG_COMPRESSED frames.
      def read_frame(io)
        header = read_exact(io, HEADER_SIZE)
        type, length = header.unpack("CV")
        payload = length > 0 ? read_exact(io, length) : "".b

        if type == MSG_COMPRESSED
          orig_type, decompressed = decompress_frame(payload)
          return [orig_type, decompressed]
        end

        [type, payload]
      end

      # Decode a RESULT_SET response payload.
      # Returns { columns: [{ name:, type_tag: }], rows: [[values]] }.
      def decode_result_set(payload)
        pos = 0
        col_count = payload[pos, 2].unpack1("v")
        pos += 2

        columns = col_count.times.map do
          name_len = payload.getbyte(pos)
          pos += 1
          name = payload[pos, name_len].force_encoding("UTF-8")
          pos += name_len
          type_tag = payload.getbyte(pos)
          pos += 1
          { name: name, type_tag: type_tag }
        end

        row_count = payload[pos, 4].unpack1("V")
        pos += 4
        null_bitmap_len = (col_count + 7) / 8

        rows = row_count.times.map do
          bitmap = payload[pos, null_bitmap_len].bytes
          pos += null_bitmap_len

          col_count.times.map do |c|
            byte_idx = c / 8
            bit_idx = c % 8
            is_null = byte_idx < bitmap.size && (bitmap[byte_idx] >> bit_idx) & 1 == 1

            if is_null
              nil
            else
              case columns[c][:type_tag]
              when TYPE_I64
                val = payload[pos, 8].unpack1("q<")
                pos += 8
                val
              when TYPE_F64
                val = payload[pos, 8].unpack1("E")
                pos += 8
                val
              when TYPE_BOOL
                val = payload.getbyte(pos) != 0
                pos += 1
                val
              when TYPE_TEXT
                len = payload[pos, 2].unpack1("v")
                pos += 2
                val = payload[pos, len].force_encoding("UTF-8")
                pos += len
                val
              when TYPE_BYTES
                len = payload[pos, 2].unpack1("v")
                pos += 2
                val = payload[pos, len].b
                pos += len
                val
              else
                len = payload[pos, 2].unpack1("v")
                pos += 2
                val = payload[pos, len].force_encoding("UTF-8")
                pos += len
                val
              end
            end
          end
        end

        { columns: columns, rows: rows }
      end

      # Decode an OK response payload.
      # Returns { rows_affected:, tag: }.
      def decode_ok(payload)
        rows_affected = payload[0, 8].unpack1("q<")
        tag_len = payload.getbyte(8)
        tag = payload[9, tag_len].force_encoding("UTF-8")
        { rows_affected: rows_affected, tag: tag }
      end

      # Decode an ERROR response payload.
      # Returns { sql_state:, message: }.
      def decode_error(payload)
        sql_state = payload[0, 5].force_encoding("ASCII")
        msg_len = payload[5, 2].unpack1("v")
        message = payload[7, msg_len].force_encoding("UTF-8")
        { sql_state: sql_state, message: message }
      end

      private

      def read_exact(io, count)
        buf = "".b
        while buf.bytesize < count
          chunk = io.read(count - buf.bytesize)
          raise PyroSQL::ConnectionError, "Connection closed by server" if chunk.nil? || chunk.empty?
          buf << chunk
        end
        buf
      end
    end
  end
end

File: go/codec.go
package pyrosql

import (
	"bytes"
	"encoding/binary"
	"errors"
	"fmt"
	"io"
	"math"

	"github.com/pierrec/lz4/v4"
)

// Message types (client -> server)
const (
	MsgQuery      byte = 0x01
	MsgPrepare    byte = 0x02
	MsgExecute    byte = 0x03
	MsgClose      byte = 0x04
	MsgPing       byte = 0x05
	MsgAuth       byte = 0x06
	MsgCompressed byte = 0x10
	MsgQuit       byte = 0xFF
)

// Capability flags for LZ4 compression negotiation.
const (
	CapLZ4               byte = 0x01
	compressionThreshold      = 8 * 1024
)

// Response types (server -> client)
const (
	RespResultSet byte = 0x01
	RespOK        byte = 0x02
	RespError     byte = 0x03
	RespPong      byte = 0x04
	RespReady     byte = 0x05
)

// Value type tags in result sets
const (
	TypeNull  byte = 0
	TypeI64   byte = 1
	TypeF64   byte = 2
	TypeText  byte = 3
	TypeBool  byte = 4
	TypeBytes byte = 5
)

const headerSize = 5

// frame builds a PWire frame: 1-byte type + 4-byte LE length + payload.
func frame(msgType byte, payload []byte) []byte {
	buf := make([]byte, headerSize+len(payload))
	buf[0] = msgType
	binary.LittleEndian.PutUint32(buf[1:5], uint32(len(payload)))
	copy(buf[5:], payload)
	return buf
}

func encodeQuery(sql string) []byte {
	return frame(MsgQuery, []byte(sql))
}

func encodePrepare(sql string) []byte {
	return frame(MsgPrepare, []byte(sql))
}

func encodeExecute(handle uint32, params []string) []byte {
	payload := make([]byte, 4+2)
	binary.LittleEndian.PutUint32(payload[0:4], handle)
	binary.LittleEndian.PutUint16(payload[4:6], uint16(len(params)))
	for _, p := range params {
		b := []byte(p)
		lenBuf := make([]byte, 2)
		binary.LittleEndian.PutUint16(lenBuf, uint16(len(b)))
		payload = append(payload, lenBuf...)
		payload = append(payload, b...)
	}
	return frame(MsgExecute, payload)
}

func encodeClose(handle uint32) []byte {
	payload := make([]byte, 4)
	binary.LittleEndian.PutUint32(payload, handle)
	return frame(MsgClose, payload)
}

func encodePing() []byte {
	return frame(MsgPing, nil)
}

func encodeAuth(user, password string) []byte {
	payload := make([]byte, 0, 2+len(user)+len(password))
	payload = append(payload, byte(len(user)))
	payload = append(payload, []byte(user)...)
	payload = append(payload, byte(len(password)))
	payload = append(payload, []byte(password)...)
	return frame(MsgAuth, payload)
}

func encodeQuit() []byte {
	return frame(MsgQuit, nil)
}

// encodeAuthWithCaps builds an AUTH frame that includes a capability byte.
func encodeAuthWithCaps(user, password string, caps byte) []byte {
	payload := make([]byte, 0, 2+len(user)+len(password)+1)
	payload = append(payload, byte(len(user)))
	payload = append(payload, []byte(user)...)
	payload = append(payload, byte(len(password)))
	payload = append(payload, []byte(password)...)
	payload = append(payload, caps)
	return frame(MsgAuth, payload)
}

// compressFrame compresses a frame payload if it exceeds the threshold.
// Returns a MSG_COMPRESSED frame or the original frame if compression is not beneficial.
func compressFrame(msgType byte, payload []byte) []byte {
	if len(payload) <= compressionThreshold {
		return frame(msgType, payload)
	}

	var buf bytes.Buffer
	w := lz4.NewWriter(&buf)
	_, err := w.Write(payload)
	if err != nil {
		return frame(msgType, payload)
	}
	if err := w.Close(); err != nil {
		return frame(msgType, payload)
	}
	compressed := buf.Bytes()

	// Check ratio: only use compression if it saves space.
	ratio := float64(len(compressed)) / float64(len(payload))
	if ratio > 0.9 {
		return frame(msgType, payload)
	}

	// Build compressed frame inner payload:
	// [original_type: u8][uncompressed_length: u32 LE][lz4_data]
	inner := make([]byte, 1+4+len(compressed))
	inner[0] = msgType
	binary.LittleEndian.PutUint32(inner[1:5], uint32(len(payload)))
	copy(inner[5:], compressed)
	return frame(MsgCompressed, inner)
}

// decompressFrame decompresses a MSG_COMPRESSED frame payload.
// Returns (original_type, decompressed_payload, error).
func decompressFrame(payload []byte) (byte, []byte, error) {
	if len(payload) < 5 {
		return 0, nil, errors.New("pwire: compressed payload too short")
	}
	originalType := payload[0]
	uncompressedLen := binary.LittleEndian.Uint32(payload[1:5])
	lz4Data := payload[5:]

	decompressed := make([]byte, 0, uncompressedLen)
	r := lz4.NewReader(bytes.NewReader(lz4Data))
	buf := make([]byte, 4096)
	for {
		n, err := r.Read(buf)
		if n > 0 {
			decompressed = append(decompressed, buf[:n]...)
		}
		if err == io.EOF {
			break
		}
		if err != nil {
			return 0, nil, fmt.Errorf("pwire: lz4 decompress: %w", err)
		}
	}

	return originalType, decompressed, nil
}

// readFrame reads a full PWire frame from a reader.
// Returns (type, payload, error).
// If the frame is MSG_COMPRESSED, it is transparently decompressed.
func readFrame(r io.Reader) (byte, []byte, error) {
	header := make([]byte, headerSize)
	if _, err := io.ReadFull(r, header); err != nil {
		return 0, nil, fmt.Errorf("pwire: read header: %w", err)
	}
	msgType := header[0]
	length := binary.LittleEndian.Uint32(header[1:5])

	if length == 0 {
		return msgType, nil, nil
	}

	payload := make([]byte, length)
	if _, err := io.ReadFull(r, payload); err != nil {
		return 0, nil, fmt.Errorf("pwire: read payload: %w", err)
	}

	// Transparently decompress MSG_COMPRESSED frames.
	if msgType == MsgCompressed {
		origType, decompressed, err := decompressFrame(payload)
		if err != nil {
			return 0, nil, err
		}
		return origType, decompressed, nil
	}

	return msgType, payload, nil
}

// Column describes a column in a result set.
type Column struct {
	Name    string
	TypeTag byte
}

// ResultSet holds a decoded RESULT_SET response.
type ResultSet struct {
	Columns []Column
	Rows    [][]interface{}
}

// OKResult holds a decoded OK response.
type OKResult struct {
	RowsAffected int64
	Tag          string
}

// PyroError represents a server-side error.
type PyroError struct {
	SQLState string
	Message  string
}

func (e *PyroError) Error() string {
	return fmt.Sprintf("pyrosql [%s]: %s", e.SQLState, e.Message)
}

func decodeResultSet(payload []byte) (*ResultSet, error) {
	if len(payload) < 2 {
		return nil, errors.New("pwire: malformed result set")
	}
	colCount := int(binary.LittleEndian.Uint16(payload[0:2]))
	pos := 2

	columns := make([]Column, colCount)
	for i := 0; i < colCount; i++ {
		if pos >= len(payload) {
			return nil, errors.New("pwire: unexpected end in column definitions")
		}
		nameLen := int(payload[pos])
		pos++
		if pos+nameLen+1 > len(payload) {
			return nil, errors.New("pwire: column name overflow")
		}
		name := string(payload[pos : pos+nameLen])
		pos += nameLen
		typeTag := payload[pos]
		pos++
		columns[i] = Column{Name: name, TypeTag: typeTag}
	}

	if pos+4 > len(payload) {
		return nil, errors.New("pwire: missing row count")
	}
	rowCount := int(binary.LittleEndian.Uint32(payload[pos : pos+4]))
	pos += 4

	nullBitmapLen := (colCount + 7) / 8
	rows := make([][]interface{}, 0, rowCount)

	for r := 0; r < rowCount; r++ {
		if pos+nullBitmapLen > len(payload) {
			return nil, errors.New("pwire: missing null bitmap")
		}
		bitmap := payload[pos : pos+nullBitmapLen]
		pos += nullBitmapLen

		row := make([]interface{}, colCount)
		for c := 0; c < colCount; c++ {
			byteIdx := c / 8
			bitIdx := uint(c % 8)
			isNull := byteIdx < len(bitmap) && (bitmap[byteIdx]>>bitIdx)&1 == 1

			if isNull {
				row[c] = nil
				continue
			}

			switch columns[c].TypeTag {
			case TypeI64:
				if pos+8 > len(payload) {
					return nil, errors.New("pwire: i64 overflow")
				}
				val := int64(binary.LittleEndian.Uint64(payload[pos : pos+8]))
				pos += 8
				row[c] = val

			case TypeF64:
				if pos+8 > len(payload) {
					return nil, errors.New("pwire: f64 overflow")
				}
				bits := binary.LittleEndian.Uint64(payload[pos : pos+8])
				val := math.Float64frombits(bits)
				pos += 8
				row[c] = val

			case TypeBool:
				if pos >= len(payload) {
					return nil, errors.New("pwire: bool overflow")
				}
				row[c] = payload[pos] != 0
				pos++

			case TypeText:
				if pos+2 > len(payload) {
					return nil, errors.New("pwire: text length overflow")
				}
				l := int(binary.LittleEndian.Uint16(payload[pos : pos+2]))
				pos += 2
				if pos+l > len(payload) {
					return nil, errors.New("pwire: text data overflow")
				}
				row[c] = string(payload[pos : pos+l])
				pos += l

			case TypeBytes:
				if pos+2 > len(payload) {
					return nil, errors.New("pwire: bytes length overflow")
				}
				l := int(binary.LittleEndian.Uint16(payload[pos : pos+2]))
				pos += 2
				if pos+l > len(payload) {
					return nil, errors.New("pwire: bytes data overflow")
				}
				b := make([]byte, l)
				copy(b, payload[pos:pos+l])
				row[c] = b
				pos += l

			default:
				// Treat unknown types as bytes (same as TEXT/BYTES wire encoding)
				if pos+2 > len(payload) {
					return nil, errors.New("pwire: unknown type length overflow")
				}
				l := int(binary.LittleEndian.Uint16(payload[pos : pos+2]))
				pos += 2
				if pos+l > len(payload) {
					return nil, errors.New("pwire: unknown type data overflow")
				}
				row[c] = string(payload[pos : pos+l])
				pos += l
			}
		}
		rows = append(rows, row)
	}

	return &ResultSet{Columns: columns, Rows: rows}, nil
}

func decodeOK(payload []byte) (*OKResult, error) {
	if len(payload) < 9 {
		return nil, errors.New("pwire: malformed OK response")
	}
	rowsAffected := int64(binary.LittleEndian.Uint64(payload[0:8]))
	tagLen := int(payload[8])
	if 9+tagLen > len(payload) {
		return nil, errors.New("pwire: OK tag overflow")
	}
	tag := string(payload[9 : 9+tagLen])
	return &OKResult{RowsAffected: rowsAffected, Tag: tag}, nil
}

func decodeError(payload []byte) (*PyroError, error) {
	if len(payload) < 7 {
		return nil, errors.New("pwire: malformed ERROR response")
	}
	sqlstate := string(payload[0:5])
	msgLen := int(binary.LittleEndian.Uint16(payload[5:7]))
	if 7+msgLen > len(payload) {
		return nil, errors.New("pwire: error message overflow")
	}
	msg := string(payload[7 : 7+msgLen])
	return &PyroError{SQLState: sqlstate, Message: msg}, nil
}
